Задания 4 урока

slv177 opened the pull request.

Commit:
Селектор графиков возвращает элемент с вложенной строкой function String() { [native code] }

Не проходит тест на высоту графиков.

Пустая функция update.

slv177 committed:
Передаем параметры в конструктор.

slv177 committed:
Селектор графиков возвращает элемент с вложенной строкой function String() { [native code] }

Не проходит тест на высоту графиков.

Пустая функция update.

slv177 committed:
Не проходит тест на высоту графиков.

Пустая функция update.

slv177 committed:
Не проходит тест на высоту графиков.

Пустая функция update.

slv177 committed:
Проходит тесты.
Неоптимальный код, дублируются целые блоки.

slv177 committed:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	04-oop-basic-intro-to-dom/1-column-chart/index.js

slv177 committed:
Проходит тесты, но нуждается в переработке.

jsru commented:
Добавляю преподавателя (@dosandk) для код-ревью.

slv177 committed:
Merge branch 'master' of https://github.com/js-tasks-ru/js-20210309-2_shambazov

slv177 committed:
TODO: убрать повторяющийся код
не проходит предпоследний тест

slv177 committed:
не проходит предпоследний тест

slv177 committed:
Проходит все тесты.

slv177 committed:
Отображается таблица с товарами.

dosandk commented on 04-oop-basic-intro-to-dom/1-column-chart/index.js:
```diff
@@ -1,3 +1,125 @@
 export default class ColumnChart {
+  constructor(options) {
```
напомню что лучше использовать значения аргументов по умолчанию, как мы разбирали на занятии

dosandk commented on 04-oop-basic-intro-to-dom/1-column-chart/index.js:
```diff
@@ -1,3 +1,125 @@
+    this.link = currentOptions.link || '';
+    this.value = currentOptions.value || 0;
+    this.chartHeight = 50;
+    this.render(options) ;
```
почему передается объект `options` в метод `render` хотя до этого все значения были сохранены внутри полей класса?

dosandk commented on 04-oop-basic-intro-to-dom/1-column-chart/index.js:
```diff
@@ -1,3 +1,125 @@
+  render(params) {
+    const element = document.createElement('div'); // (*)
+
+    if (this.data.length) {
```
попробуйте избавиться от `if...else` структуры и оставить только один шаблон

dosandk commented on 04-oop-basic-intro-to-dom/1-column-chart/index.js:
```diff
@@ -1,3 +1,125 @@
+
+    this.element = element.firstElementChild;
+
+    let labelLabel = this.element.querySelector(".column-chart__title");
```
проверьте, здесь точно `let`?

dosandk commented on 04-oop-basic-intro-to-dom/1-column-chart/index.js:
```diff
@@ -1,3 +1,125 @@
+    let labelLabel = this.element.querySelector(".column-chart__title");
+    labelLabel.innerHTML = '<a href="/' + this.link + '"class="column-chart__link"> Total ' + this.label + '</a>';
+
+    let valueLabel = this.element.querySelector(".column-chart__header");
```
проверьте, здесь точно `let`?

dosandk commented on 04-oop-basic-intro-to-dom/1-column-chart/index.js:
```diff
@@ -1,3 +1,125 @@
+    let valueLabel = this.element.querySelector(".column-chart__header");
+    valueLabel.innerHTML = this.value;
+
+    function getColumnProps(data) {
```
почему данная функция объявлена дважды?

slv177 replied:
Не догадался запускать через this. 
Исправлено.

dosandk commented on 04-oop-basic-intro-to-dom/1-column-chart/index.js:
```diff
@@ -1,3 +1,125 @@
+      return result;
+    }
+
+    function getColumnProps(data) {
```
почему данная функция объявлена дважды?

slv177 replied:
Не догадался запускать через this. 
Исправлено.

dosandk commented on 04-oop-basic-intro-to-dom/1-column-chart/index.js:
```diff
@@ -1,3 +1,125 @@
+  }
+
+  update(params) {
+    var graphbar = document.querySelector('.column-chart__chart');
```
помните на первом занятии мы договаривались о том что мы не будет использовать `var`?

dosandk commented on 04-oop-basic-intro-to-dom/1-column-chart/index.js:
```diff
@@ -1,3 +1,125 @@
+    var graphbar = document.querySelector('.column-chart__chart');
+
+    if (graphbar.hasChildNodes) {
+      console.log(params);
```
не оставляйте отладочных вызовов `console` в PR

dosandk commented on 04-oop-basic-intro-to-dom/1-column-chart/index.js:
```diff
@@ -1,3 +1,125 @@
+
+
+    function drawChartGraph (dataForGraph) {
+      let result = String();
```
дайте пожалуйста ответ на вопрос: почему здесь строка объявлена через `String`, а не через литерал?

slv177 replied:
Когда писал код, написал let result =  new String(); потом IDE предложила убрать new.
обычно пишу let result = "";
Лучше использовать второй вариант?

slv177 replied:
Исправлено.

dosandk replied:
да, следует использовать просто литерал строки `result = ''`

dosandk commented on 05-dom-document-loading/1-notification/index.js:
```diff
@@ -1,3 +1,65 @@
 export default class NotificationMessage {
+  constructor(message, options = {}) {
+    this.text = message || '';
+    this.duration = options['duration'] || 5000;
```
пожалуйста напомните в чем разница между обращением к свойствам объекта через точку и через квадратные скобки

slv177 replied:
Через точку мы указываем название свойства, через скобки можно так же указать через вычисляемую переменную. Я, видимо по привычке из питона пишу, как обращался бы к словарю с указанием ключа.
Правильнее через точку?

slv177 replied:
Переделал конструктор целиком.

dosandk replied:
good 👍 
да, через точку нужно обращаться к тем свойствам которые заранее известны в объекте. А через квадратные скобки к тем свойствам которые должны быть взяты например из переменной

dosandk commented on 05-dom-document-loading/1-notification/index.js:
```diff
@@ -1,3 +1,65 @@
 export default class NotificationMessage {
+  constructor(message, options = {}) {
```
пожалуйста воспользуйтесь аргументами по умолчанию

dosandk commented on 05-dom-document-loading/1-notification/index.js:
```diff
@@ -1,3 +1,65 @@
+      this.element.classList.add('error');
+      break;
+    default:
+      console.log("unknown type");
```
не оставляйте отладочных `console.log`

dosandk commented on 05-dom-document-loading/1-notification/index.js:
```diff
@@ -1,3 +1,65 @@
+  show (elementToDisplay) {
+    if (!NotificationMessage.isDisplay) {
+      if (elementToDisplay) {
+        let newElement = elementToDisplay.append(this.element);
```
проверьте, здесь точно `let`?

dosandk commented on 05-dom-document-loading/1-notification/index.js:
```diff
@@ -1,3 +1,65 @@
+  }
+
+
+  show (elementToDisplay) {
```
подумайте, может ли значение по умолчанию для аргумента `elementToDisplay` упростить логику внутри метода?

slv177 replied:
И точно! разобрался )

dosandk reviewed (changes requested):
пожалуйста дайте ответы на вопросы, внесите правки согласно комментариям.
Для решения SortableTable, пожалуйста проверьте использование `let/const` и устраните дубликаты кода. Также проверьте чем отличается обращение к свойствам объекта через точку и через квадратные скобки и поправьте решения

slv177 committed:
Работает сортировка.

slv177 committed:
Исправлена часть замечаний.

slv177 committed:
table header

slv177 committed:
Проходит тесты.Конструктор переделан на рекомендованый, но возможно неоптимален.

slv177 committed:
Конструктор переделан на рекомендованый, но возможно неоптимален. Не знаю, насколько правильно обращаться к аргументам функции через arguments[0].

slv177 committed:
Все работает, но не проходит ни один тест. TypeError: Cannot read property '0' of undefined

slv177 committed:
Исправлены замечания.

slv177 committed:
Переработан метод show().

dosandk commented on 04-oop-basic-intro-to-dom/1-column-chart/index.js:
```diff
@@ -1,3 +1,91 @@
+  }
+
+  update(params) {
+    let graphbar = document.querySelector('.column-chart__chart');
```
проверьте здесь точно `let`?

dosandk commented on 04-oop-basic-intro-to-dom/1-column-chart/index.js:
```diff
@@ -1,3 +1,91 @@
+    }
+
+    const labelLabel = this.element.querySelector(".column-chart__title");
+    labelLabel.innerHTML = '<a href="/' + this.link + '"class="column-chart__link"> Total ' + this.label + '</a>';
```
пожалуйста используйте шаблонную строку (обратные кавычки) вместо конкатенации строк через `+`

dosandk commented on 04-oop-basic-intro-to-dom/1-column-chart/index.js:
```diff
@@ -1,3 +1,91 @@
 export default class ColumnChart {
+  constructor({
```
good 👍

dosandk commented on 04-oop-basic-intro-to-dom/1-column-chart/index.js:
```diff
@@ -1,3 +1,91 @@
+  render() {
+    const element = document.createElement('div'); // (*)
+
+    element.innerHTML = `
```
good 👍

dosandk commented on 04-oop-basic-intro-to-dom/1-column-chart/index.js:
```diff
@@ -1,3 +1,91 @@
+    const valueLabel = this.element.querySelector(".column-chart__header");
+    valueLabel.innerHTML = this.value;
+
+    let chartGraph = this.element.querySelector('.column-chart__chart');
```
проверьте, здесь точно `let`?

dosandk commented on 05-dom-document-loading/1-notification/index.js:
```diff
@@ -1,3 +1,65 @@
+    duration = 1000,
+    type = ''
+  } = {}) {
+    if (arguments[1]) {
```
от `arguments` следует избавиться. Ведь у нас уже объявлены аргументы `text`, `duration`, `type` с ними и следует работать

dosandk replied:
не используйте `arguments`

slv177 committed:
Конкатенация через обратные кавычки.

dosandk commented on 05-dom-document-loading/1-notification/index.js:
```diff
@@ -1,3 +1,65 @@
+        <div class="inner-wrapper">
+          <div class="notification-header">success</div>
+          <div class="notification-body">
+            ` + this.text + `
```
так делать нельзя, для подстановки значения в шаблонную строку следует использовать конструкцию:
```
...
<div class="notification-body">
  ${this.text}
</div>
...
```

dosandk commented on 05-dom-document-loading/2-sortable-table-v1/index.js:
```diff
@@ -1,4 +1,130 @@
+    header = [],
+    data = {},
+  } = {}) {
+    this.header = arguments[0];
```
от `arguments` нужно избавиться

dosandk commented on 05-dom-document-loading/2-sortable-table-v1/index.js:
```diff
@@ -1,4 +1,130 @@
+  }
+
+  render(dataToRender) {
+    let element = document.createElement('div');
```
проверьте весь метод на использование `let` и `const`

dosandk commented on 05-dom-document-loading/2-sortable-table-v1/index.js:
```diff
@@ -1,4 +1,130 @@
+    let body = document.querySelector(".sortable-table__body")
+    while (tableBodyRowCollection.length > 0)
+    {
+      element.appendChild(tableBodyRowCollection[0])
```
используйте `append`, https://dev.to/ibn_abubakre/append-vs-appendchild-a4m

dosandk commented on 05-dom-document-loading/2-sortable-table-v1/index.js:
```diff
@@ -1,4 +1,130 @@
+      element.appendChild(tableBodyRowCollection[0])
+    }
+
+    console.log("finally", element);
```
не должно быть отладочных консолей в пулл-реквестах - это дурной тон

slv177 replied:
Этот файл еще не вносил в PR, еще не закончил с ним работу. Или он попал?

dosandk commented on 05-dom-document-loading/2-sortable-table-v1/index.js:
```diff
@@ -1,4 +1,130 @@
+    this.render(this.data);
+  }
+
+  sortStrings(field = 'title', order = 'asc') {
```
из этого метода нужно убрать дубликаты кода. Посмотрите занятие на котором мы разбирали решение `sortStrings`

slv177 replied:
Этот файл еще не вносил в PR, еще не закончил с ним работу. Или он попал?

dosandk commented on 05-dom-document-loading/2-sortable-table-v1/index.js:
```diff
@@ -1,4 +1,130 @@
+                    <div class="sortable-table__cell">
+                        <img class="sortable-table-image" alt="Image" src="${item['images'][0]['url']}">
+                    </div>
+                    <div class="sortable-table__cell"> ${item['title']}1</div>
```
что будет если мы захотим отобразить данные в которых не будет колонки `price` или `quantity`?

slv177 replied:
Над заданием еще работаю.

dosandk replied:
ок

dosandk commented on 05-dom-document-loading/2-sortable-table-v1/index.js:
```diff
@@ -1,4 +1,130 @@
 
+  get tableHeaderRow() {
+    let result = '';
+    // result += this.header.forEach(element =>
```
не должно быть в PR закомментированного кода

dosandk reviewed (changes requested):
необходимо внести правки согласно комментариям

slv177 committed:
Работает.
Не проходят тесты кроме первых двух.
Надо избавиться от arguments в конструкторе.

slv177 committed:
Работает.
Не проходят тесты кроме первых двух.
Надо избавиться от arguments в конструкторе.

slv177 committed:
Работает.
Не проходят тесты кроме первых двух.
Надо избавиться от arguments в конструкторе.

slv177 committed:
Revert "Работает. Не проходят тесты кроме первых двух. Надо избавиться от arguments в конструкторе."

This reverts commit c04847ea

slv177 committed:
Merge remote-tracking branch 'origin/master'

slv177 committed:
Merge remote-tracking branch 'origin/master'

slv177 committed:
Merge remote-tracking branch 'origin/master'

slv177 committed:
Удалены неиспользуемые участки кода.

slv177 committed:
Удалены неиспользуемые участки кода.

slv177 committed:
Merge remote-tracking branch 'origin/master'

slv177 committed:
Удалены неиспользуемые участки кода.

slv177 committed:
исправлен конструктор.

slv177 committed:
Добавлен subElements, но тесты после 3-го не проходят - соединяет значения из первых столбцов.

slv177 committed:
Merge branch 'master' of https://github.com/js-tasks-ru/js-20210309-2_shambazov

dosandk commented on 05-dom-document-loading/1-notification/index.js:
```diff
@@ -1,3 +1,65 @@
+    this.render();
+  }
+
+  static isDisplay = false;
```
статические свойства обычно размещают над `constructor`, при чтении класса это помогает понять какие поля являются статическими

dosandk commented on 05-dom-document-loading/2-sortable-table-v1/index.js:
```diff
@@ -1,4 +1,151 @@
+    element.appendChild(tableHeaderRow);
+    let tableHeaderRowCollection = tableHeaderRow.children;
+    let header = document.querySelector(".sortable-table__header")
+    while (tableHeaderRowCollection.length > 0)
```
достаточно просто проверки:
```
while (tableHeaderRowCollection.length)
```

dosandk commented on 05-dom-document-loading/2-sortable-table-v1/index.js:
```diff
@@ -1,4 +1,151 @@
+    let header = document.querySelector(".sortable-table__header")
+    while (tableHeaderRowCollection.length > 0)
+    {
+      header.appendChild(tableHeaderRowCollection[0])
```
пожалуйста используйте `append` 
https://dev.to/ibn_abubakre/append-vs-appendchild-a4m

dosandk commented on 05-dom-document-loading/2-sortable-table-v1/index.js:
```diff
@@ -1,4 +1,151 @@
+  get tableHeaderRow() {
+    let result = '';
+
+    for (const item of this.header) {
```
пожалуйста используйте `map` вместо конкатенации строк

slv177 replied:
Заменено.

dosandk commented on 05-dom-document-loading/2-sortable-table-v1/index.js:
```diff
@@ -1,4 +1,151 @@
+    let tableBody = document.createElement('div');
 
+    for (const rowData of dataToDisplay) {
+      let row = document.createElement('a');
```
проверьте, здесь точно `let`?

dosandk commented on 05-dom-document-loading/2-sortable-table-v1/index.js:
```diff
@@ -1,4 +1,151 @@
+  }
+
+  getSubElements(element) {
+    // const elements = element.querySelectorAll('[data-element]');
```
пожалуйста не оставляйте закомментированный код

slv177 replied:
Это рабочий файл, этот код оставлен, чтоб показать, что при получении субэлементов двумя способами результат тот же.

slv177 committed:
до фикса

slv177 committed:
до фикса

slv177 committed:
осталось исправить сортировку и сделать map

slv177 committed:
осталось исправить сортировку и сделать map

dosandk commented on 06-events-practice/2-tooltip/index.js:
```diff
@@ -1,5 +1,69 @@
 class Tooltip {
+  constructor() {
+    // тут какие то объявления со стрелками
```
объявления со стрелками до `constructor`

dosandk commented on 06-events-practice/2-tooltip/index.js:
```diff
@@ -1,5 +1,69 @@
+    }
+
+  addEventListeners(){
+    const tooltip = document.querySelector("[data-tooltip]");
```
нельзя привязываться к конкретному элементу, элементы на странице могут добавляться и удаляться динамически

dosandk commented on 06-events-practice/2-tooltip/index.js:
```diff
@@ -1,5 +1,69 @@
+  addEventListeners(){
+    const tooltip = document.querySelector("[data-tooltip]");
+
+    tooltip.addEventListener('pointerover', event => {
```
обработчики следует добавлять на `document`

dosandk commented on 06-events-practice/2-tooltip/index.js:
```diff
@@ -1,5 +1,69 @@
+    const tooltip = document.querySelector("[data-tooltip]");
+
+    tooltip.addEventListener('pointerover', event => {
+      console.log("tooltipFoo", 'pointerover', event.clientX, event.clientY);
```
все отладочные `console` нужно удалить из PR

dosandk commented on 06-events-practice/2-tooltip/index.js:
```diff
@@ -1,5 +1,69 @@
+
+    tooltip.addEventListener('pointerover', event => {
+      console.log("tooltipFoo", 'pointerover', event.clientX, event.clientY);
+      let target = event.target;
```
проверьте использования `let` и `const`

dosandk commented on 06-events-practice/2-tooltip/index.js:
```diff
@@ -1,5 +1,69 @@
+      showTooltop(event);
+    });
+
+    function showTooltop(event) {
```
почему бы не вынести данную функцию в метод класса?

dosandk commented on 06-events-practice/2-tooltip/index.js:
```diff
@@ -1,5 +1,69 @@
+      let message = document.createElement('div');
+      message.classList.add('tooltip');
+      message.innerHTML = target.dataset.tooltip;
+      message.style.left = event.clientX + 10 + 'px';
```
10 неплохо было бы положить в переменную с осмысленным именем

dosandk commented on 06-events-practice/2-tooltip/index.js:
```diff
@@ -1,5 +1,69 @@
+  removeEventListeners(){
+    const tooltip = document.querySelector("[data-tooltip]");
+
+    tooltip.removeEventListener('pointerout', event => {
```
так обработчик события не снимается, должна быть использована та же функция/метод. Пересмотрите пожалуйста последнее занятие

slv177 committed:
Сортировка без повторения строк (по признаку).
реализован map вместо конкатенации.

slv177 committed:
Работает и проходят тесты, но надо доработать навешивание и снятие обработчиков.

dosandk commented on 05-dom-document-loading/2-sortable-table-v1/index.js:
```diff
@@ -53,13 +53,9 @@ export default class SortableTable {
-    for (const item of this.header) {
-      result += `<div class="sortable-table__cell" data-id= ${item.id} data-sortable="false"><span> ${item.title} </span></div>`;
-    }
+    return this.header.map(item => `<div class="sortable-table__cell" data-id=${item.id} data-sortable="false"><span>${item.title}</span></div>`).join("");
```
good 👍

dosandk commented on 06-events-practice/2-tooltip/index.js:
```diff
@@ -1,69 +1,64 @@
 class Tooltip {
-  constructor() {
-    // тут какие то объявления со стрелками
+  showTooltip = (event) => {
```
good 👍

dosandk commented on 06-events-practice/2-tooltip/index.js:
```diff
@@ -1,69 +1,64 @@
+    }
+    );
+
+    document.addEventListener('pointermove', event => {
```
данный обработчик лучше навесить при наведении на элемент

dosandk reviewed (approved):
good 👍 посмотрите комментарий
